File: learning/transformer-js/core/heads.js
/**
 * 分类头模块
 * 实现各种下游任务的分类器和预测头
 * 
 * @author shihom_wu
 * @version 1.0.0
 */

const { Linear, Dropout } = require('./layers');
const { softmax } = require('./math-utils');

/**
 * 序列分类头
 * 用于文本分类、情感分析等任务
 */
class SequenceClassificationHead {
  /**
   * 构造函数
   * 
   * @param {Object} config - 配置对象
   * @param {number} config.hidden_size - 隐藏层大小
   * @param {number} config.num_labels - 标签数量
   * @param {number} config.dropout - Dropout概率，默认0.1
   */
  constructor(config) {
    this.hiddenSize = config.hidden_size;
    this.numLabels = config.num_labels;
    this.dropout = new Dropout(config.dropout || 0.1);
    this.classifier = new Linear(this.hiddenSize, this.numLabels);
  }

  /**
   * 前向传播
   * 
   * @param {Array<Array<number>>} hiddenStates - 隐藏状态 [seqLen, hiddenSize]
   * @param {number} poolingStrategy - 池化策略: 0=CLS token, 1=mean pooling, 2=max pooling
   * @returns {Array<number>} 分类logits [numLabels]
   */
  forward(hiddenStates, poolingStrategy = 0) {
    let pooledOutput;

    switch (poolingStrategy) {
      case 0: // CLS token pooling
        pooledOutput = hiddenStates[0]; // 使用第一个token ([CLS])
        break;
      case 1: // Mean pooling
        pooledOutput = this._meanPooling(hiddenStates);
        break;
      case 2: // Max pooling
        pooledOutput = this._maxPooling(hiddenStates);
        break;
      default:
        throw new Error(`不支持的池化策略: ${poolingStrategy}`);
    }

    // 应用dropout和分类器
    const droppedOutput = this.dropout.forward(pooledOutput);
    return this.classifier.forward(droppedOutput);
  }

  /**
   * 平均池化
   * 
   * @param {Array<Array<number>>} hiddenStates - 隐藏状态
   * @returns {Array<number>} 池化后的向量
   */
  _meanPooling(hiddenStates) {
    const seqLen = hiddenStates.length;
    const hiddenSize = hiddenStates[0].length;
    const pooled = new Array(hiddenSize).fill(0);

    for (let i = 0; i < seqLen; i++) {
      for (let j = 0; j < hiddenSize; j++) {
        pooled[j] += hiddenStates[i][j];
      }
    }

    return pooled.map(val => val / seqLen);
  }

  /**
   * 最大池化
   * 
   * @param {Array<Array<number>>} hiddenStates - 隐藏状态
   * @returns {Array<number>} 池化后的向量
   */
  _maxPooling(hiddenStates) {
    const hiddenSize = hiddenStates[0].length;
    const pooled = new Array(hiddenSize).fill(-Infinity);

    for (let i = 0; i < hiddenStates.length; i++) {
      for (let j = 0; j < hiddenSize; j++) {
        pooled[j] = Math.max(pooled[j], hiddenStates[i][j]);
      }
    }

    return pooled;
  }

  /**
   * 获取参数数量
   * 
   * @returns {number} 参数总数
   */
  getParameterCount() {
    return this.classifier.getParameterCount();
  }

  /**
   * 设置训练模式
   * 
   * @param {boolean} training - 是否为训练模式
   */
  setTraining(training) {
    this.dropout.setTraining(training);
  }
}

/**
 * Token分类头
 * 用于命名实体识别、词性标注等token级别任务
 */
class TokenClassificationHead {
  /**
   * 构造函数
   * 
   * @param {Object} config - 配置对象
   * @param {number} config.hidden_size - 隐藏层大小
   * @param {number} config.num_labels - 标签数量
   * @param {number} config.dropout - Dropout概率，默认0.1
   */
  constructor(config) {
    this.hiddenSize = config.hidden_size;
    this.numLabels = config.num_labels;
    this.dropout = new Dropout(config.dropout || 0.1);
    this.classifier = new Linear(this.hiddenSize, this.numLabels);
  }

  /**
   * 前向传播
   * 
   * @param {Array<Array<number>>} hiddenStates - 隐藏状态 [seqLen, hiddenSize]
   * @returns {Array<Array<number>>} 每个token的分类logits [seqLen, numLabels]
   */
  forward(hiddenStates) {
    const results = [];

    for (let i = 0; i < hiddenStates.length; i++) {
      const droppedOutput = this.dropout.forward(hiddenStates[i]);
      const logits = this.classifier.forward(droppedOutput);
      results.push(logits);
    }

    return results;
  }

  /**
   * 获取参数数量
   * 
   * @returns {number} 参数总数
   */
  getParameterCount() {
    return this.classifier.getParameterCount();
  }

  /**
   * 设置训练模式
   * 
   * @param {boolean} training - 是否为训练模式
   */
  setTraining(training) {
    this.dropout.setTraining(training);
  }
}

/**
 * 语言模型头
 * 用于MLM、CLM等语言建模任务
 */
class LanguageModelingHead {
  /**
   * 构造函数
   * 
   * @param {Object} config - 配置对象
   * @param {number} config.hidden_size - 隐藏层大小
   * @param {number} config.vocab_size - 词汇表大小
   * @param {boolean} config.tie_word_embeddings - 是否共享词嵌入权重
   */
  constructor(config) {
    this.hiddenSize = config.hidden_size;
    this.vocabSize = config.vocab_size;
    this.tieWordEmbeddings = config.tie_word_embeddings || false;
    
    if (!this.tieWordEmbeddings) {
      this.lmHead = new Linear(this.hiddenSize, this.vocabSize, false);
    }
  }

  /**
   * 前向传播
   * 
   * @param {Array<Array<number>>} hiddenStates - 隐藏状态 [seqLen, hiddenSize]
   * @param {Array<Array<number>>|null} embeddingWeights - 词嵌入权重（如果共享权重）
   * @returns {Array<Array<number>>} 词汇表上的logits [seqLen, vocabSize]
   */
  forward(hiddenStates, embeddingWeights = null) {
    if (this.tieWordEmbeddings && embeddingWeights) {
      // 使用共享的嵌入权重
      return this._computeWithSharedWeights(hiddenStates, embeddingWeights);
    } else {
      // 使用独立的语言模型头
      const results = [];
      for (let i = 0; i < hiddenStates.length; i++) {
        const logits = this.lmHead.forward(hiddenStates[i]);
        results.push(logits);
      }
      return results;
    }
  }

  /**
   * 使用共享权重计算logits
   * 
   * @param {Array<Array<number>>} hiddenStates - 隐藏状态
   * @param {Array<Array<number>>} embeddingWeights - 嵌入权重 [vocabSize, hiddenSize]
   * @returns {Array<Array<number>>} logits
   */
  _computeWithSharedWeights(hiddenStates, embeddingWeights) {
    const results = [];
    
    for (let i = 0; i < hiddenStates.length; i++) {
      const logits = [];
      const hidden = hiddenStates[i];
      
      // 计算 hidden @ embedding_weights.T
      for (let j = 0; j < embeddingWeights.length; j++) {
        let score = 0;
        for (let k = 0; k < hidden.length; k++) {
          score += hidden[k] * embeddingWeights[j][k];
        }
        logits.push(score);
      }
      
      results.push(logits);
    }
    
    return results;
  }

  /**
   * 获取参数数量
   * 
   * @returns {number} 参数总数
   */
  getParameterCount() {
    return this.tieWordEmbeddings ? 0 : this.lmHead.getParameterCount();
  }

  /**
   * 设置训练模式
   * 
   * @param {boolean} training - 是否为训练模式
   */
  setTraining(training) {
    if (!this.tieWordEmbeddings) {
      // 语言模型头没有dropout等需要设置训练模式的组件
    }
  }
}

// 导出所有分类头
module.exports = {
  SequenceClassificationHead,
  TokenClassificationHead,
  LanguageModelingHead
};

File: learning/transformer-python/transformer_pytorch/core/heads.py
"""
分类头模块
实现各种下游任务的分类器和预测头

作者: shihom_wu
版本: 1.0.0
"""

import torch
import torch.nn as nn
import torch.nn.functional as F
from typing import Optional, Union, Tuple
from torch import Tensor


class SequenceClassificationHead(nn.Module):
    """
    序列分类头
    用于文本分类、情感分析等任务
    """
    
    def __init__(
        self,
        hidden_size: int,
        num_labels: int,
        dropout: float = 0.1,
        pooling_strategy: str = 'cls'
    ):
        """
        初始化序列分类头
        
        Args:
            hidden_size: 隐藏层大小
            num_labels: 标签数量
            dropout: Dropout概率
            pooling_strategy: 池化策略 ('cls', 'mean', 'max')
        """
        super().__init__()
        self.hidden_size = hidden_size
        self.num_labels = num_labels
        self.pooling_strategy = pooling_strategy
        
        self.dropout = nn.Dropout(dropout)
        self.classifier = nn.Linear(hidden_size, num_labels)
        
        # 初始化权重
        nn.init.normal_(self.classifier.weight, std=0.02)
        nn.init.zeros_(self.classifier.bias)
    
    def forward(
        self, 
        hidden_states: Tensor,
        attention_mask: Optional[Tensor] = None
    ) -> Tensor:
        """
        前向传播
        
        Args:
            hidden_states: 隐藏状态 [batch_size, seq_len, hidden_size]
            attention_mask: 注意力掩码 [batch_size, seq_len]
            
        Returns:
            分类logits [batch_size, num_labels]
        """
        # 根据池化策略获取序列表示
        if self.pooling_strategy == 'cls':
            # 使用第一个token ([CLS])
            pooled_output = hidden_states[:, 0]
        elif self.pooling_strategy == 'mean':
            # 平均池化
            pooled_output = self._mean_pooling(hidden_states, attention_mask)
        elif self.pooling_strategy == 'max':
            # 最大池化
            pooled_output = self._max_pooling(hidden_states, attention_mask)
        else:
            raise ValueError(f"不支持的池化策略: {self.pooling_strategy}")
        
        # 应用dropout和分类器
        pooled_output = self.dropout(pooled_output)
        logits = self.classifier(pooled_output)
        
        return logits
    
    def _mean_pooling(
        self, 
        hidden_states: Tensor, 
        attention_mask: Optional[Tensor] = None
    ) -> Tensor:
        """平均池化"""
        if attention_mask is None:
            return hidden_states.mean(dim=1)
        
        # 考虑注意力掩码的平均池化
        mask_expanded = attention_mask.unsqueeze(-1).expand_as(hidden_states).float()
        sum_embeddings = torch.sum(hidden_states * mask_expanded, dim=1)
        sum_mask = torch.clamp(mask_expanded.sum(dim=1), min=1e-9)
        
        return sum_embeddings / sum_mask
    
    def _max_pooling(
        self, 
        hidden_states: Tensor, 
        attention_mask: Optional[Tensor] = None
    ) -> Tensor:
        """最大池化"""
        if attention_mask is None:
            return hidden_states.max(dim=1)[0]
        
        # 考虑注意力掩码的最大池化
        mask_expanded = attention_mask.unsqueeze(-1).expand_as(hidden_states)
        hidden_states = hidden_states.masked_fill(~mask_expanded.bool(), -1e9)
        
        return hidden_states.max(dim=1)[0]


class TokenClassificationHead(nn.Module):
    """
    Token分类头
    用于命名实体识别、词性标注等token级别任务
    """
    
    def __init__(
        self,
        hidden_size: int,
        num_labels: int,
        dropout: float = 0.1
    ):
        """
        初始化Token分类头
        
        Args:
            hidden_size: 隐藏层大小
            num_labels: 标签数量
            dropout: Dropout概率
        """
        super().__init__()
        self.hidden_size = hidden_size
        self.num_labels = num_labels
        
        self.dropout = nn.Dropout(dropout)
        self.classifier = nn.Linear(hidden_size, num_labels)
        
        # 初始化权重
        nn.init.normal_(self.classifier.weight, std=0.02)
        nn.init.zeros_(self.classifier.bias)
    
    def forward(self, hidden_states: Tensor) -> Tensor:
        """
        前向传播
        
        Args:
            hidden_states: 隐藏状态 [batch_size, seq_len, hidden_size]
            
        Returns:
            每个token的分类logits [batch_size, seq_len, num_labels]
        """
        hidden_states = self.dropout(hidden_states)
        logits = self.classifier(hidden_states)
        
        return logits


class LanguageModelingHead(nn.Module):
    """
    语言模型头
    用于MLM、CLM等语言建模任务
    """
    
    def __init__(
        self,
        hidden_size: int,
        vocab_size: int,
        tie_word_embeddings: bool = False,
        bias: bool = True
    ):
        """
        初始化语言模型头
        
        Args:
            hidden_size: 隐藏层大小
            vocab_size: 词汇表大小
            tie_word_embeddings: 是否共享词嵌入权重
            bias: 是否使用偏置
        """
        super().__init__()
        self.hidden_size = hidden_size
        self.vocab_size = vocab_size
        self.tie_word_embeddings = tie_word_embeddings
        
        if not tie_word_embeddings:
            self.lm_head = nn.Linear(hidden_size, vocab_size, bias=bias)
            # 初始化权重
            nn.init.normal_(self.lm_head.weight, std=0.02)
            if bias:
                nn.init.zeros_(self.lm_head.bias)
    
    def forward(
        self, 
        hidden_states: Tensor,
        embedding_weights: Optional[Tensor] = None
    ) -> Tensor:
        """
        前向传播
        
        Args:
            hidden_states: 隐藏状态 [batch_size, seq_len, hidden_size]
            embedding_weights: 词嵌入权重 [vocab_size, hidden_size]（如果共享权重）
            
        Returns:
            词汇表上的logits [batch_size, seq_len, vocab_size]
        """
        if self.tie_word_embeddings and embedding_weights is not None:
            # 使用共享的嵌入权重: hidden @ embedding_weights.T
            logits = torch.matmul(hidden_states, embedding_weights.T)
        else:
            # 使用独立的语言模型头
            logits = self.lm_head(hidden_states)
        
        return logits


class MultipleChoiceHead(nn.Module):
    """
    多选题分类头
    用于多选题回答等任务
    """
    
    def __init__(
        self,
        hidden_size: int,
        dropout: float = 0.1
    ):
        """
        初始化多选题分类头
        
        Args:
            hidden_size: 隐藏层大小
            dropout: Dropout概率
        """
        super().__init__()
        self.hidden_size = hidden_size
        
        self.dropout = nn.Dropout(dropout)
        self.classifier = nn.Linear(hidden_size, 1)
        
        # 初始化权重
        nn.init.normal_(self.classifier.weight, std=0.02)
        nn.init.zeros_(self.classifier.bias)
    
    def forward(self, hidden_states: Tensor) -> Tensor:
        """
        前向传播
        
        Args:
            hidden_states: 隐藏状态 [batch_size, num_choices, hidden_size]
            
        Returns:
            每个选择的分数 [batch_size, num_choices]
        """
        hidden_states = self.dropout(hidden_states)
        logits = self.classifier(hidden_states).squeeze(-1)
        
        return logits


class QuestionAnsweringHead(nn.Module):
    """
    问答任务头
    用于抽取式问答任务，预测答案的开始和结束位置
    """
    
    def __init__(
        self,
        hidden_size: int,
        dropout: float = 0.1
    ):
        """
        初始化问答任务头
        
        Args:
            hidden_size: 隐藏层大小
            dropout: Dropout概率
        """
        super().__init__()
        self.hidden_size = hidden_size
        
        self.dropout = nn.Dropout(dropout)
        self.qa_outputs = nn.Linear(hidden_size, 2)  # 开始和结束位置
        
        # 初始化权重
        nn.init.normal_(self.qa_outputs.weight, std=0.02)
        nn.init.zeros_(self.qa_outputs.bias)
    
    def forward(self, hidden_states: Tensor) -> Tuple[Tensor, Tensor]:
        """
        前向传播
        
        Args:
            hidden_states: 隐藏状态 [batch_size, seq_len, hidden_size]
            
        Returns:
            开始位置logits和结束位置logits
        """
        hidden_states = self.dropout(hidden_states)
        logits = self.qa_outputs(hidden_states)
        
        start_logits, end_logits = logits.split(1, dim=-1)
        start_logits = start_logits.squeeze(-1)
        end_logits = end_logits.squeeze(-1)
        
        return start_logits, end_logits


# 导出所有分类头
__all__ = [
    'SequenceClassificationHead',
    'TokenClassificationHead',
    'LanguageModelingHead',
    'MultipleChoiceHead',
    'QuestionAnsweringHead'
]
